crates/router/src/core/api_locking.rs

```diff
@@ -0,0 +1,190 @@
+    if is_locking_enabled_for_merchant && is_locking_enabled_on_api {
+        let expiry_in_seconds = 60; // get it from redis
+        let delay_between_retries_in_milli_seconds = 100; // get it from redis
+        let retries = 1; // get from redis based on should_retry, if not present in redis default 1?
+        lock_resource(
+            state,
+            locking_input,
+            expiry_in_seconds,
+            retries,
+            delay_between_retries_in_milli_seconds,
+        )
+        .await
+    } else {
+        logger::info!(
+            "Resource Locking not Enabled for merchant_id: {} and api: {}",
+            locking_input.merchant_id,
+            api_identifier.to_owned()
+        );
+        Ok(LockStatus::NotEnabled)
+    }
+}
+
+#[instrument(skip(state))]
+pub async fn lock_resource(
+    state: &routes::AppState,
+    locking_input: LockingInput,
+    expiry_in_seconds: u64,
+    retries: u32,
+    delay_between_retries_in_milli_seconds: u64,
+) -> RouterResult<LockStatus> {
+    acquire_lock_on_resource_in_redis(
+        state,
+        locking_input.get_redis_key_for_locks(),
+        true,
+        expiry_in_seconds,
+        delay_between_retries_in_milli_seconds,
+        retries,
+    )
+    .await
+}
+
+#[instrument(skip(state))]
+pub async fn acquire_lock_on_resource_in_redis(
+    state: &routes::AppState,
+    key: &str,
+    value: bool,
+    expiry_in_seconds: u64,
+    delay_between_retries_in_milli_seconds: u64,
+    mut retries: u32,
+) -> RouterResult<LockStatus> {
+    let redis_conn = state.store.clone().get_redis_conn();
+
+    for _retry in 0..retries {
+        let is_lock_acquired = redis_conn
+            .set_key_if_not_exists_with_expiry(
+                key,
+                value,
+                Some(
+                    expiry_in_seconds
+                        .try_into()
+                        .into_report()
+                        .change_context(errors::ApiErrorResponse::InternalServerError)?,
+                ),
+            )
+            .await;
+
+        match is_lock_acquired {
+            Ok(redis::SetnxReply::KeySet) => {
+                return Ok(LockStatus::Acquired(key.to_owned()));
+            }
+            Ok(redis::SetnxReply::KeyNotSet) => {
+                logger::error!("Lock not acquired, retrying");
+                actix_time::sleep(tokio::time::Duration::from_millis(
+                    delay_between_retries_in_milli_seconds * 1000,
+                ))
+                .await;
+            }
+            Err(error) => {
+                logger::error!(error=%error.current_context(), "Error while locking");
+                actix_time::sleep(tokio::time::Duration::from_millis(
+                    delay_between_retries_in_milli_seconds * 1000,
+                ))
+                .await;
+            }
+        }
+    }
+
+    Ok(LockStatus::AcquireFailedRetriesExhausted(key.to_owned()))
+}
+
+#[instrument(skip(state))]
+pub async fn release_lock(
```

Comment:
change the status to just have acquired and released. And check requestId as authentication to release the lock